### PR TITLE
Adds the incandescent husk, an evolution of the shocker zombie that moves very slowly and emits a field of lightning

### PR DESCRIPTION
**Pull request by ghost**
```SUMMARY: Content "Adds the incandescent husk, an evolution of the shocker zombie that moves slowly and emits a lightning cloud"```

### Short version

This adds the incandescent husk, a new rare evolution of the shocker zombie. The husk moves very slowly, and emits a small cloud of lightning around it like how smoker zombies emit smoke. EMPing the husk will briefly disable its electric field, but be wary of it during lightning storms...

### Long version

This idea occurred to me after playing Prey and fighting an enemy called the [Voltaic Phantom](https://steamuserimages-a.akamaihd.net/ugc/834700315685045075/728C0A720D2EDB6596C3C53FCD620E793266DFDA/?interpolation=lanczos-none&output-format=jpeg&output-quality=95&fit=inside%7C1024%3A576&composite-to=*,*%7C1024%3A576&background-color=black) that emits lightning around it and moves very slowly. The incandescent husk evolves from the shocker zombie, and serves a different function. Its speed of 50 means that a walking player can outrun it, and it has a very weak melee attack, at 1d4 + 4 electric damage, but its cloak of lightning engulfs everything near it, and can spell death for a cocky survivor.

This electric field also has unique effects. First, it emits a constant buzzing sound that can be heard from up to fifteen tiles away. It also ignites any flammable liquid that's adjacent to the husk, and subject anything next to it to electromagnetic pulse. Finally, whenever it spends its turn next to a gas pump, that gas pump has a 1-in-4 chance to explode from the lightning cloaking it.

EMP blasts will nullify the electrical field for 3 minutes (30 turns), giving you a chance to get in and take it out with melee. Note that it's still electrically charged, so hitting it will still shock you if your weapon is conductive or you're using your fists.

I think the husk is an interesting addition both because of its extreme danger if played around foolishly, as well as its potential to be used as an advantage if it isn't. Its slow speed means that, if it can be kited well, it can electrocute other zombies near it, potentially removing them from the fight. It also doesn't have a ranged lightning attack like its precursor.

As these only spawn after a year, players should have plenty of ways to fight it, such as ranged weapons, the anti-shock CBM, or a suit that nullifies electricity damage. Direct melee combat is difficult due to the constant shock field causing speed reductions, damage, and pain, but can still be pulled off if high enough damage is dealt per attack.

Fluffwise, I like to imagine the husk as someone who was heavily augmented with bionics before the Cataclysm. Evolved from shocker zombies, the evolution damaged their bionics in some way, causing the electric field that surrounds them - as well as their lightning bolt - to combine into a cloud of deadly electric discharge, crawling greedily along the ground as its host moves.

Stat line:
* 200 HP
* 50 speed
* 50 difficulty (dark red name)
* 3 melee skill
* 1d4 melee damage + 4 electric damage + 2 cut
* 3 bash armor, 5 cut armor
* Emits lightning slightly less thickly than a smoker zombie emits smoke
* EMP blasts disable the field
* Dissection may yield bionics

### CODE REVIEW

File: data/json/monsters.json [hunk 1]
```diff
@@ -3076,6 +3076,52 @@
...
+      "REVIVES",
+      "BONES"
+    ],
+    "upgrades": { "half_life": 28, "into": "mon_zombie_nullfield" }
+  },
+  {
+    "id": "mon_zombie_nullfield",
+    "type": "MONSTER",
+    "name": "null zombie",
+    "description": "This once-human is visible only as a glowing white silhouette that you have to squint to see, cloaked in a crackling field of lightning that pulses like a beating heart.  It walks slowly and deliberately, the thunderstorm surrounding it eagerly jumping to anything conductive within its grasp.",
```
Comment (0be1isk) [inline data/json/monsters.json]: This once-human?

Comment (Xhuis) [reply data/json/monsters.json]: Once a human, now a zombie.

**Comment by 0be1isk** [comment]
I understand that part, just think you should clarify, or else it may sound like a grammatical error

**Comment by Lorith** [comment]
I like the idea, but at the same time think it is a bit off, the other zombies with cbms could potentially power them due to delays between usage, but this seems like it would be difficult even for something brute-sized with muscles similar to an electric eel to power continually.  Perhaps add a shocker hulk and branch the shocker brute into both that and this one, with an increase in health and possibly armor?  That could also solve the issue of shocker brutes vanishing into short-ranged enemies.

**Comment by Xhuis** [comment]
@nexusmrsep Shocker brutes are described as being charged enough that they emit audible humming while being near them, so I wanted to evolve them into that. Furthermore, I deliberately wanted to avoid having to death with both null zombies and shocker brutes simultaneously, as it'd be hellish.

@Lorith I ran some testing with 50x spawn rates a year in, and found that null zombies were the outliers, with more shocker brutes being present. I am also very against shocker hulks, as shocker brutes are already meant to serve the niche that a hulk would - very strong, fast, and dangerous. My current thoughts for power is a still-running microreactor, torsion ratchets, or simply the blob inside it evolving to generate a strong natural current.

That said, it actually occurs to me reading your post that having it alternate between "emitting a large electrical cloud" and "emitting nothing" would be an interesting design. Were that the chase, I'd probably modify null zombies to be an evolution of normal shockers, and reduce their difficulty to light-red name instead of dark-red.

**Comment by Lorith** [comment]
Having them set to a few turns between bursts would work very well in terms of powering the cbms, just put it as an overloaded offensive defense cbm or malfunctioning chain lightning cbm.  Could also potentially have a range of times between switches so it can do longer or shorter periods between bursts, and occasionally stronger bursts than others, to make even racing in as soon as it stops risky without proper preparation.

File: data/json/monsters.json (hunk 1 above)
Comment (0be1isk) [reply data/json/monsters.json]: I understand that part, just think you should clarify, or else it may sound like a grammatical error

**Comment by MT-Arnoldussen** [comment]
By the time these arrive, wouldn't they be much easier to deal with than shocker brutes?

**Comment by Xhuis** [comment]
With anti-electrical gear or equipment, yes; without it, no. I'm still mulling over the design, so things are subject to change.

**Comment by DracoGriffin** [comment]
This seems like a perfect synergy evolution / monster with the zombie burner.

What's the possibility of allowing the electrical fields sparking the leaking gasoline/napalm/propellant? I believe that was one of the ideas brought up in #23729.

Hopefully they would both be sufficiently rare enough to prevent a ton of issues, but somewhat more than a "once in a lifetime" [player-wise] thing that could be an interesting "event" in-game.

Also, I agree on upgrading from shocker zombies rather than brutes, but that might just be the name. Null juggernaut (now that we have skeletal) or something would invoke the feeling of evolving beyond the shocker brute.

A few other name suggestions: "[Spark/Gleamer] [Zombie/Brute/Juggernaut]"

**Comment by Xhuis** [comment]
Made some changes in this commit.
* Added a new flag, `ELECTRIC_FIELD`, and gave it to the null zombie. It causes flammable liquids such as gasoline adjacent to the null zombie to catch fire. I may expand it to affect electronics, water, etc. if I can figure out how.
* Null zombies now evolve from shockers instead of shocker brutes, so you can live in electric hell if you really want to.

**Comment by Xhuis** [comment]
Now named the incandescent husk.

The electric field has been expanded. Now, the husk emits a `zzzZZZzzzZZZzzzZZZ` sound every five turns. Everything adjacent to the husk will undergo EMP effects due to the constant discharge, and any gas pumps have a 1-in-4 chance to explode for each turn that the discharge field engulfs them.

File: src/monmove.cpp [hunk 2]
```diff
@@ -1132,14 +1132,37 @@ bool monster::move_to( const tripoint &p, bool force, const float stagger_adjust
...
+            const auto t = g->m.ter( zap );
+            if( t == ter_str_id( "t_gas_pump" ) || t == ter_str_id( "t_gas_pump_a" ) ) {
+                if( one_in( 4 ) ) {
+                    g->explosion( p, 40, 0.8, true );
+                    if( player_sees ) {
+                        add_msg( m_warning, _( "The %s explodes in a fiery inferno!" ), g->m.tername( zap ).c_str() );
+                    }
+                } else {
+                    if( player_sees ) {
+                        add_msg( m_warning, _( "Lightning from %s engulfs the %s!" ), name().c_str(),
```
Comment (Night-Pryanik) [inline src/monmove.cpp]: Please use positional arguments.

**Comment by kevingranade** [comment]
On Thu, Aug 2, 2018, 11:51 PM Ashe Higgs <notifications@github.com> wrote:

> Furthermore, I deliberately wanted to avoid having to death with both null
> zombies and shocker brutes simultaneously, as it'd be hellish.
>
Sounds like this became a non-issue here, but evolution should never
eliminate a kind of zombie from the population.

>

**Comment by Xhuis** [comment]
In testing, the sound appears to be quiet enough that it's only really hearable from quite close by. It's mainly intended both for atmosphere and so that the player is aware of the danger if the husk is inside of a building, but I'm unsure if I'll keep it.

**Comment by YakumoChen** [comment]
Zombies tend to also bunch up around shockers in the first place (due to being lit up) and I still feel like it will end up just being a killbox for zombies with no need to have effort on the player's part

**Comment by Xhuis** [comment]
That's true that it may end up causing it to kill other zombies during night or in dark areas. I'm not sure of a way to solve that, save for perhaps making it flicker on and off.

**Comment by kevingranade** [comment]
Sounds worth investigating, "zombies bunch up around lightning zombies"

**Comment by Xhuis** [comment]
I believe that's more of a quirk of their AI, as they tend to move towards light when it's dark.

**Comment by Xhuis** [comment]
Big update in this commit.
* Adds the `insulated` effect, which provides immunity to electricity.
* EMP blasts now apply the `emp` effect for 3 minutes, which doesn't do anything on its own but serves to track that an EMP blast has hit you recently.
* Incandescent husks with the `emp` effect won't emit their electrical field until the effect expires.
* Husks apply the `insulated` effect to friendly monsters within 5 tiles for 1 minute, stacking up to 5 minutes.

This compensates for the removal of a counterplay option (you can't kite zombies into the husk when it makes them immune to its own lightning) by adding another in the form of EMP. The removal of kiting into it is mainly to make it less trivial to fight, and to make it a large threat even in crowds.

The insulation of nearby monsters substantially increases its difficulty to kill, so I will be monitoring closely when it's added in to see if it requires a nerf.

Unless there's anything else needed, this should be ready to go.

**Comment by Xhuis** [comment]
Wait no stop I found a bug

**Comment by Xhuis** [comment]
Migrated the field to `process_turn` instead of `moveto` because it was not properly interacting with the world around it when the monster wasn't moving.

I also made insulation capped at 5 turns when given by the husk so that lightning damage is still viable against them through chain lightning and other methods.

**Comment by Night-Pryanik** [comment]
The insulation works on zombies and doesn't work on the player character? That doesn't make sense.

**Comment by Xhuis** [comment]
Zombie masters and necromancers both interact in unique ways with zombies and not players. Unless my code knowledge is mistaken, if the player uses zombie pheromones and turns the husk friendly, it will insulate the player.

It could make sense in a lot of ways, If it's a damaged Chain Lightning CBM that produces the lightning field, it could simply direct the lightning away from the zombies. It could surround them in harmless lightning that thus combines with incoming electricity, or emit a pheromone that tells zombies around it to develop grounding properties (as the blob is already showable to evolve on a whim.)

**Comment by Night-Pryanik** [comment]
Damaged bionic that shocker doesn't even control selectively protects one sort of beings and damages other one?

**Comment by Xhuis** [comment]
Obtusively speaking, that would technically be the case. The shocker zombie and brute are both capable of directing their electrical discharge at specific targets; allowing the incandescent husk to emit a field around it in a large cone and avoiding others of its pack makes sense to me.

If you think there's a better way to handle the issue, I'd be happy to hear it, but I feel this to be the smoothest way of ensuring that it doesn't serve as a barbecue for any zombies that wander too close to it.

**Comment by Night-Pryanik** [comment]
Maybe there's no problem at all. You monster emits a lightning field around itself? Ok. Zombies are attracted by the light it generates? Ok. Zombies are dumb and not insulated to avoid the field? Yes. Zombies are dying in mass because of that? Fine. That seems logical to me.

If you really want to compensate that, you could just make your monster emits its field less frequently.

**Comment by Xhuis** [comment]
Having a zombie that kills other zombies by being near them is fine, but having one that lures other zombies to it because of a quirk in zombie pathfinding isn't. I feel like having it kill zombies around it makes it *less* dangerous as an evolved form.

I feel like it would be fairly simple to direct the electrical discharge, given that the Chain Lightning CBM allows players to do exactly that, and even shocker zombies can direct their electricity even when it surrounds them enough that striking them zaps you. Were its insulation mechanic removed, there are a lot of other things I could do with it, but I dislike the idea of something being able to destroy an entire horde by existing as a part of it.

**Comment by kevingranade** [comment]
tl;dr The husk as a loose cannon is a good concept, I don't want to lose that over a pathing bug.  The problem here is the attraction effect, not the damage.

The insulation effect has several problems.
On it's face having an AoE attack that ignores friendly enemies (which is the observable end result of the immunity) is far more sophisticated than targeting the player.  Even targeted attacks from spitters and shockers hit other zombies as well, so there's no precedent for this level of sophistication.
The effect itself is problematic, what's the mechanism? How are players informed it's happening? Is there a countermeasure?  
The immunity effect also has side effects, like preventing other electrical attacks from working on zombies near the Husk, both from other shockers and from the player.
From a game point of view it's very heavy handed and surprising, everything about this monster tells you it should be dangerous to it's surroundings, but then you notice that it's only causing "indiscriminate damage" to certain things, that's surprising and *will be perceived as unfair*.

> I feel like having it kill zombies around it makes it less dangerous as an evolved form.
> I dislike the idea of something being able to destroy an entire horde by existing as a part of it.

We don't need more masses of zombies, the player encounters that all the time, what we need more of is the kind of single very dangerous enemy the Husk represents.
We aren't trying to make the zombies win, we're trying to present interesting challenges to the player.

**Comment by Lorith** [comment]
Agree with everything kevin said here, but the idea of an insulating zombie is also interesting.  Perhaps split that off into another rare evolution from a shocker where it consumes any electrical fields nearby, give it high HP and armor, with no glow effect.  Maybe also place some in the necropolis.  The two together would be basically just strong zombies, but then you could have the husk generating fields on its own and the insulator stopping chain lightning from erasing hordes.

As for attracting zombies, that would have to be due to sound, as light does not directly attract zombies.

**Comment by Xhuis** [comment]
Sounds were indeed causing the issue, so the constant vrummm been removed.

Made some updates in this commit.

Removed the insulating effect. I also reduced the husk's armor, but increased the electric damage it deals on a melee attack to make it more threatening to armored players. In addition, it no longer smashes furniture (due to being frail beneath the lightning), and no longer stumbles like other zombies (as it walks slowly and deliberately.)

Finally, it now functions as a lightning rod. During lightning storms, it may be struck by lightning, which has a few effects through a new effect called `supercharged` that lasts for twelve hours.
* Speed increases by 50
* Radius of the electric cloud *doubles*
* Constantly makes loud "VMMMMMM" sounds; while this will likely attract more zombies, it's much more dangerous as an individual now so this is acceptable to me

This isn't meant to be a static modifier to its perceived difficulty, but rather to create an interesting occurrence and give the player a pause for thought as to whether or not it's worth the risk to attack it even when it's supercharged.

**Comment by Mecares** [comment]
As a non natural english speaker the name is not very intuitive to me, which of course may be caused by my lack of english skills but maybe something like overloaded shocker zombie, supercharged or overcharge shocker zombie are decent naming alternatives?
